File: src/LinkList/LinkList.java
package LinkList;

import sun.awt.image.ImageWatched;

import java.util.List;

//https://www.geeksforgeeks.org/implementing-a-linked-list-in-java-using-class/
public class LinkList {

    Node head; //head for list

    class Node
    {
        int data;
        Node next;

        Node(int item)
        {
            data = item; //creating a node for every item
            next=null;
        }
    }
    /* Function to insert in linklist*/
    LinkList insert(LinkList list, int item)
    {
        Node new_node = new Node(item);
        new_node.next=null;

        if(list.head == null)
        {
            list.head=new_node;    //assigning first element in the list

        }
        else
        {
            Node last = list.head;
            while (last.next != null)
            {
                last=last.next;  //traversing to the last node
            }
            last.next = new_node;


        }
        System.out.println(item+" inserted successfully");
        return list;

    }

    LinkList beginingInsert(LinkList list, int data)
    {
        Node new_node = new Node(data);

        if(list.head == null)
        {
            list.head = new_node;
        }
        else
        {
            new_node.next=head;
            head=new_node;
        }
        System.out.println(data+" inserted at the beginning");
        return list;
    }

    LinkList insertAtPosition(LinkList list, int data, int position)
    {
        Node newNode = new Node(data);
        int count=0;
        Node prev = null;
        Node temp = list.head;

        if(position==0)
        {
            newNode.next=list.head;
            head=newNode;
            System.out.println(data+" inserted successfully at position "+position);
        }
        else{


        while(temp!=null && count<position)
        {
            prev = temp;
            temp=temp.next;
            count++;
        }
        if(count == position)
        {
            prev.next=newNode;
            newNode.next=temp;
            System.out.println(data+" inserted successfully at position "+position);
        }
        else
        {
            System.out.println("Data not inserted");
        }
        }
        return list;
    }

    LinkList delete(LinkList list, int item)
    {
        Node node = list.head;
        Node prev = null;
        if(node.data == item)
        {
            list.head=node.next;
            System.out.println(item+" deleted");

        }
        else
        {

            while (node!=null && node.data!=item)
            {
                prev=node;
                node=node.next;
            }
            if(node!=null)
            {
                prev.next=node.next;
                System.out.println(item+" deleted");
            }
            else
            {
                System.out.println("item not found");
            }
        }
        return list;
    }

    //to delete a specific position
    LinkList deleteAtPosition(LinkList list, int position)
    {
        Node node = list.head;
        Node prev = null;
        int count=0;
        if(position == 0 && node !=null)
        {
            System.out.println("Element"+node.data+" at position "+position+" removed.");
            //head to be removed
            list.head = node.next;

        }
        else
        {
            while(node.next!=null & count<position)
            {
                prev=node;
                node=node.next;
                count++;
            }
            if(count==position)
            {
                System.out.println("Element"+node.data+" at position "+position+" removed.");
                prev.next=node.next;
            }
            else
            {
                System.out.println("Element at position "+position+" not found.");
            }
        }
        return list;
    }

    LinkList deleteAtBegin(LinkList list)
    {
        Node node = list.head;
        if(list.head!=null)
        {
            System.out.println(node.data+" deleted from the beginning");
            list.head=node.next;

        }
        else
        {
            System.out.println("empty list");
        }
        return list;
    }

    LinkList deleteAtEnd(LinkList list)
    {
        Node temp = list.head;
        Node prev = null;
        if(list.head==null)
        {
            System.out.println("empty list");
        }
        else
        {
            while(temp.next!=null)
            {
                prev = temp;
                temp=temp.next;
            }

                System.out.println(temp.data+" deleted from the end");
                prev.next=null;

        }
        return list;
    }

    void search(LinkList list, int item)
    {
        Node node = list.head;
        int count =0;
        boolean flag =false;
        if(list.head == null)
        {
            System.out.println("List is empty");
        }
        else
        {
            while(node.next != null)
            {
                if(node.data == item)
                {
                    flag=true;
                    System.out.println(node.data+" found at position "+count);
                    break;
                }
                count++;
                node=node.next;
            }
        }
        if(flag == false)
        {
            System.out.println("Item not found");
        }
    }
    void print(LinkList list)
    {
        Node currentNode = list.head;
        while (currentNode != null)
        {
            System.out.print(currentNode.data+" -> ");
            currentNode=currentNode.next;
        }
    }


}

File: src/LinkList/linklist.c
#include<stdio.h>
#include<stdlib.h>

struct  node{
    int data;
    struct node *next;
};
struct node *head;

void beginsert()
{
    struct node *ptr, *temp;
    int item;

    ptr=(struct node *) malloc(sizeof(struct node));
    printf("Enter an item to insert at beginning: ");
    scanf("%d",&item);
    ptr->data=item;
    if(ptr==NULL)
    {
        printf("Underflow!\n");
    }
    else
    {
        ptr->next=head;
        head=ptr;
        printf("Item inserted at the beginning.\n");
    }
    

}
void insert()
{
    struct node *ptr, *temp;
    int item;

    ptr= (struct node *) malloc (sizeof(struct node));
    printf("Enter an item to insert at the end: ");
    scanf("%d",&item);
    ptr->data=item;
    if(head==NULL)
    {
        ptr->next = NULL;
        head = ptr;
        printf("Item inserted\n");
    }
    else
    {
        temp=head;
        while(temp->next!=NULL)
        {
            temp=temp->next;
        }
        temp->next=ptr;
        ptr->next=NULL;
        printf("Item inserted\n");
    }
}
void insertAtPosition()
{
    int item,position,count=0,flag=0;
    
    struct node *ptr,*temp,*prev;
    ptr= (struct node *) malloc(sizeof(struct node));
    
    printf("Enter the item to be inserted: ");
    scanf("%d",&item);
    printf("\nEnter the position: ");
    scanf("%d",&position);

    if(ptr==NULL)
    {
        printf("Underflow");
    }
    else{
        ptr->data=item;
        if(position==0)
        {
            ptr->next=head;
            head=ptr;
        }
        else
        {   
            temp = head;
            while(count<position)
            {
                if(temp == NULL)
                {
                    printf("Unable to insert\n");
                    flag=1;
                    break;
                }
                else{
                prev=temp;
                temp=temp->next;
                count=count+1;
                }
            }
            if(flag == 0)
        {   prev->next=ptr;
            ptr->next=temp;
            printf("Item Inserted\n");
        }    
        }
    }


}


void deleteAnElement()
{
    struct node *ptr,*prev,*temp;
    int item;
    ptr = (struct node *) malloc(sizeof(struct node));

    printf("Enter an Item to delete: ");
    scanf("%d",&item);

    if(head==NULL)
    {
        printf("Unable to delete");
    }
    else{
        temp=head;
        if(temp->data==item)
        {
            head=temp->next;
            printf("Item deleted\n");

        }
        else{
        while(temp->data!=item && temp!=NULL)
        {
            prev=temp;
            temp=temp->next;
        }
        if(temp->data==item)
        {
            prev->next=temp->next;
            printf("Item deleted\n");

        }
        }
    }

}

void deleteAtBegin()
{
    struct node *ptr;
    ptr=(struct node *) malloc(sizeof(struct node));
    ptr=head;
    if(head!=NULL)
    {
        head=ptr->next;
        printf("Item deleted from Beginning\n");
    }
    else
    {
        printf("Unable to delete\n");
    }
}

void deleteAtEnd()
{
    struct node *ptr,*temp;
    ptr=(struct node *) malloc(sizeof(struct node));
    ptr=head;
    if(head==NULL)
    {
        printf("unable to delete\n");
    }
    else
    {
        
        while(ptr->next!=NULL)
        {
            temp=ptr;
            ptr=ptr->next;
        }
        temp->next=NULL;
        printf("Last Element Deleted\n");

    }
}

void search()
{
    struct node *ptr;
    int item,count=0;
    ptr=(struct node *) malloc(sizeof(struct node));
    printf("Enter an element to be searched: ");
    scanf("%d",&item);
    if(head==NULL)
    {
        printf("Item not found\n");
    }
    else
    {   
        ptr=head;
        while(ptr->data!=item && ptr->next!=NULL)
        {
            ptr=ptr->next;
            count=count+1;
        }
        if(ptr->data==item)
        {
            printf("Item found at position %d\n",count);
        }
        else
        {
            printf("item not found\n");
        }
    }
}

void display()
{
    struct node *ptr;
    ptr=head;
    if(ptr==NULL)
    {
        printf("Empty LinkedList\n");
    }
    else
    {
        while(ptr != NULL)
        {
            printf("%d -> ",ptr->data);
            ptr=ptr->next;
        }
        printf("\n");
    }
}


int main()
{

 int choice =0;  
    while(choice != 9)   
    {  
        printf("\n\n*********Main Menu*********\n");  
        printf("\nChoose one option from the following list ...\n");  
        printf("\n===============================================\n");  
        printf("\n1.Insert in begining\n2.Insert at last\n3.Insert at any random location\n4.Delete from Beginning\n5.Delete from last\n6.Delete node after specified location\n7.Search for an element\n8.Show\n9.Exit\n");  
        printf("\nEnter your choice?\n");         
        scanf("\n%d",&choice);  
        switch(choice)  
        {  
            case 1:  
            beginsert();      
            break;  
            case 2:  
            insert();         
            break;  
            case 3:  
            insertAtPosition();       
            break;  
            case 4:  
            deleteAtBegin();       
            break;  
            case 5:  
            deleteAtEnd();        
            break;  
            case 6:  
            deleteAnElement();          
            break;  
            case 7:  
            search();         
            break;  
            case 8:  
            display();        
            break;  
            case 9:  
            exit(0);  
            break;  
            default:  
            printf("Please enter valid choice..");  
        }  
    }  
    return 0;

}
